rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,81 @@
+# RFC: Enabling Determinism in TensorFlow
+  
+| Status        | Proposed                                             |
+:---------------|:-----------------------------------------------------|
+| **Author(s)** | Pankaj Kanwar (Google), Reed Wanderman-Milne (Google)|
+| **Sponsor**   | Sanjoy Das (Google)                                  |
+| **Updated**   | 2021-01-31                                           |
+
+
+## Objective
+Allow users to enable determinism behavior in TensorFlow. This means if the user runs a TensorFlow program multiple times, the model outputs and weights will be the same each time. Determinism will be supported on CPUs and GPUs.
+ 
+To get deterministic behavior, users must do the following:
```

Comment:
One question here is whether Keras is in scope here as well? I believe in some conversations with @fchollet that there might be some non-determinism introduced in the Keras framework as well

Reply:
So the other source of randomness in Keras (besides variable initializers) is going to be layer calls (like dropout).

We intend these to rely on stateless ops, seeded by a `seed` argument in `call`. That seed would be either provided by a per-model stateful RNG when calling the model, or if left unspecified would be autogenerated by a global RNG.

Reply:
In support of what Reed wrote above, in my experience, all nondeterminism seen when using Keras is essentially due to underlying issues with non-keras TensorFlow.